addons/crm/static/src/js/forecast/forecast_service.js

```diff
@@ -0,0 +1,230 @@
+    }
+    /**
+     * @param {Array[]} domain
+     * @param {boolean} limited whether [this.end] moment must be used as a domain constraint to limit
+     *                          read_group results [limited === true]
+     *                          or not [limited === false]
+     * @returns {Array[]} modified domain
+     */
+    getDomain(domain, limited=undefined) {
+        if (typeof limited === 'undefined') {
+            limited = !!this.end;
+        }
+        if (!limited) {
+            return domain;
+        }
+        var extraDomain = ['|', [this.fieldName, '=', false],
+            [this.fieldName, '<', this.end.format(this.format)]];
+        if (domain.length) {
+            domain = ['&', ...domain, ...extraDomain];
+        } else {
+            domain = extraDomain;
+        }
+        return domain;
+    }
+    /**
+     * @param {Object} context
+     * @param {boolean} greedy whether fill_temporal must apply until [this.end] [greedy === true]
+     *                         or until the last group with at least one record [greedy === false]
+     * @returns {Object} modified context
+     */
+    getContext(context, greedy=undefined) {
+        if (typeof greedy === 'undefined') {
+            greedy = !!this.end;
+        }
+        context = {
+            ...context,
+            fill_temporal: {
+                fill_from: this.start.format(this.format),
+                min_groups: this.min_groups
+            }
+        };
+        if (greedy) {
+            context.fill_temporal.fill_to = moment(this.end).subtract(1, this.minGranularity).format(this.format);
+        }
+        return context;
+    }
+    /**
+     * @param {integer} min_groups minimum amount of groups to display, regardless of other constraints
+     */
+    set_min_groups(min_groups) {
+        this.min_groups = min_groups || 1;
+    }
+}
+
+/**
+ * Forecast Service
+ *
+ * This service will be used to generate or retrieve Forecast Periods
+ * 
+ * A specific forecast period configuration will always refer to the same instance
+ * unless reset is true
+ */
+ var ForecastService = AbstractService.extend({
+    /**
+     * @override
+     */
+    start() {
+        this._super.apply(...arguments);
+        this._forecastPeriods = {};
+    },
+    /**
+     * Get a forecast period according to the configuration
+     * 
+     * @param {Object} configuration
+     * @param {string} [modelName]
+     * @param {string} [fieldName]
+     * @param {string} [fieldType]
+     * @param {string} [granularity] can either be : hour, day, week, month, quarter, year
+     * @param {integer} [min_groups] optional minimal amount of desired groups
+     * @param {boolean} [reset] optional whether the forecast period should be reinstancied
+     * @returns {ForecastPeriod}
+     */
+    getPeriod({modelName, fieldName, fieldType, granularity, min_groups=4, reset=false}) {
+        if (!(modelName in this._forecastPeriods)) {
```

Comment:
This is the `res_model` from the action `crm_lead_action_forecast` (as it is loaded in the property `loadParams` of the JS model). 
It is necessary, because if 2 such `res_model` have a field with the same name, their forecast would be confused. (The service could be used by 2 distinct parts of Odoo, but the forecastPeriod instances would be the same)

Reply:
Ok. But then use real modelName (seems there is a "lead" propagated, should be "crm.lead").